addons/stock/report/report_stock_reception.py

```diff
@@ -0,0 +1,285 @@
+
+from collections import defaultdict
+
+from odoo import _, api, models
+from odoo.tools import float_compare, float_is_zero, format_date
+
+
+class ReceptionReport(models.AbstractModel):
+    _name = 'report.stock.report_reception'
+    _description = "Stock Reception Report"
+
+    @api.model
+    def _get_report_values(self, docids, data=None):
+        ''' This report is flexibly designed to work with both individual and batch pickings.
+        '''
+        docids = self.env.context.get('default_picking_ids', docids)
+        pickings = self.env['stock.picking'].search([('id', 'in', docids), ('picking_type_code', '!=', 'outgoing'), ('state', '!=', 'cancel')])
+        picking_states = pickings.mapped('state')
+        # unsupported cases
+        if not pickings:
+            msg = "No transfers selected or a delivery order selected"
+        elif 'done' in picking_states and len(set(picking_states)) > 1:
+            pickings = False
+            msg = "This report cannot be used for done and not done transfers at the same time"
+        if not pickings:
+            return {'pickings': False, 'reason': msg}
+
+        # incoming move qtys
+        product_to_qty_draft = defaultdict(float)
+        product_to_qty_to_assign = defaultdict(list)
+        product_to_total_assigned = defaultdict(list)
+
+        # to support batch pickings we need to track the total already assigned
+        assigned_moves = pickings.move_lines.mapped('move_dest_ids')
+        product_to_assigned_qty = defaultdict(float)
+        for assigned in assigned_moves:
+            product_to_assigned_qty[assigned.product_id] += assigned.product_qty
+
+        for move in pickings.move_lines:
+            if move.product_id.type != 'product':
+                continue
+            qty_already_assigned = 0
+            if move.move_dest_ids:
+                qty_already_assigned = min(product_to_assigned_qty[move.product_id], move.product_qty)
+                product_to_assigned_qty[move.product_id] -= qty_already_assigned
+            if qty_already_assigned:
+                product_to_total_assigned[move.product_id].append((move, qty_already_assigned))
+            if move.product_qty != qty_already_assigned:
+                if move.state == 'draft':
+                    product_to_qty_draft[move.product_id] += move.product_qty - qty_already_assigned
+                else:
+                    quantity_to_assign = move.product_qty
+                    if move.state == 'done':
+                        quantity_to_assign = move.product_uom._compute_quantity(move.quantity_done, move.product_id.uom_id, rounding_method='HALF-UP')
+                    product_to_qty_to_assign[move.product_id].append((quantity_to_assign - qty_already_assigned, move))
+
+        # only match for non-mto moves in same warehouse
+        warehouse = pickings[0].picking_type_id.warehouse_id
+        wh_location_ids = [loc['id'] for loc in self.env['stock.location'].search_read(
```

Comment:
equivalent to `self.env['stock.location'].search([('id', 'child_of', warehouse.view_location_id.id)]).ids`